[Model Monitoring] Manage `last_analyzed` exclusively in the controller

jond01 opened the pull request:
Use JSON files, one per endpoint.
Implements [ML-7457](https://iguazio.atlassian.net/browse/ML-7457).

Related CE config update PR:
https://github.com/mlrun/ce/pull/141

[ML-7457]: https://iguazio.atlassian.net/browse/ML-7457?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jond01 committed:
Fix `get_model_monitoring_file_target_path` signature

jond01 committed:
Improve DataItem type hints

jond01 committed:
Fix `InMemoryStore._get_item` error

jond01 committed:
Update the controller to use data store DB

jond01 committed:
Add a new type for schedules

jond01 committed:
Delete `last_analyzed` from MM stores

jond01 committed:
Catch also FileStore error

jond01 committed:
Fix an `InMemoryStore` test

jond01 committed:
Edit messages

jond01 committed:
Clean the controller from outdated code

jond01 committed:
Add a small test for `batch_dict2timedelta`

jond01 committed:
Support S3 for CE

davesh0812 commented on mlrun/model_monitoring/controller.py:
```diff
@@ -41,173 +39,174 @@
+        f"{endpoint}.json",
+    )
+
+
```
please move this function to `model_monitoring.helpers`.

davesh0812 commented on mlrun/model_monitoring/controller.py:
```diff
@@ -41,173 +39,174 @@
+            str,
+            mlrun.mlconf.get_model_monitoring_file_target_path(
+                project=project,
+                kind=mm_constants.FileTargetKind.MONITORING_SCHEDULES,
```
You are using the online target logic (in spite it is an offline target), it means that you are always  get the  `mlrun.mlconf.model_endpoint_monitoring.store_prefixes.default` configuration.
In iguazio it will work fine but in CE I didn't saw that we overwrite this configuration (in default it using v3io).

jond01 replied:
Right, I planned to update it in the CE config: https://github.com/mlrun/ce/pull/141

davesh0812 commented on mlrun/model_monitoring/controller.py:
```diff
@@ -41,173 +39,174 @@
+                    raise
+            last_analyzed = self._init_last_analyzed()
+            schedules = _Schedules({self._application: last_analyzed})
+            logger.info(
+                "The monitoring schedules file does not exist for this endpoint, creating it",
+                path=self._db.url,
```
Please add the creation of this json to be part of the model endpoint creation, and also we need to delete it when calling to `delete_model_endpoints_resources`.

davesh0812 commented on mlrun/model_monitoring/controller.py:
```diff
@@ -41,173 +39,174 @@
+
+
+# `Schedules` is a dictionary of registered application name as key and Unix timestamp as value
+_Schedules = NewType("_Schedules", dict[str, int])
```
Can we create a `Schedules` class that get the json file and have few method like `get(), update(), update_db(), delete()` and some more maybe... I just think it can be more user friendly for us. And can be used also from  MEP creation/deletion API.

jond01 replied:
I think that I get what you mean, but this is basically the `DataItem` API we use here, isn't it?
I'll provide some more abstraction to reduce the duplication between MEP creation/deletion and the controller.

Edit: I united the APIs in the `ModelMonitoringSchedulesFile` class.

davesh0812 commented on mlrun/model_monitoring/controller.py:
```diff
@@ -41,173 +39,174 @@
+                timestamp + self._step, tz=datetime.timezone.utc
+            )
+            yield _Interval(start_time, end_time)
+            self._update_last_analyzed(timestamp + self._step)
```
You don't think it is to much to update the store object on each app? Why no to update it one time per mep?
If we will have 10 apps and 100 mep we will reach to the db 1000 in each run of the controller.

jond01 replied:
The reason is to avoid data loss/duplication. These files are the persistent storage that ensures no loss upon unexpected controller failures.

Edit: Done - the mode endpoint schedules file is read once and written once per controller run.

davesh0812 reviewed (changes requested):
LookNice, I didn't finish all the  review.
Few important comments - the main one is that we have to delete/create the json when deleting/creating the model endpoint or delete it in case of project deletion.

jond01 committed:
Remove unused imports after rebase

jond01 committed:
Post-rebase fix

jond01 committed:
Move the helper file path function to helpers

jond01 committed:
Create the schedules file in `create_model_endpoint`

jond01 committed:
Use `StrEnum`

jond01 committed:
Create and delete the schedules files in an organized way

jond01 committed:
Add get folder path helper

jond01 committed:
Remove an unused import

jond01 committed:
Create the schedules file in the tests

jond01 committed:
Fix `test_helpers.py`

jond01 committed:
Add a helper get schedules DataItem helper

jond01 committed:
Add `_ModelMonitoringSchedulesFile` tests

jond01 committed:
Move `ModelMonitoringSchedulesFile` to a module

Including the tests

jond01 committed:
Move the delete schedules folder to a function

Add a test

jond01 committed:
Fix imports post-rebase

jond01 committed:
Fix context and event type hints

jond01 committed:
Controller refactor 1

jond01 committed:
Refactor part 1

jond01 committed:
Refactor part 2

jond01 committed:
Add a termination callback to the controller

jond01 committed:
Add `_should_monitor_endpoint`

Check first and last request existence

jond01 committed:
Merge remote-tracking branch 'upstream/development' into task/controller-last-analyzed-in-json

davesh0812 commented on mlrun/model_monitoring/controller.py:
```diff
@@ -45,164 +45,129 @@ class _Interval(NamedTuple):
 
-class _BatchWindowGenerator:
-    def __init__(self, batch_dict: Union[dict, str]) -> None:
+class _BatchWindowGenerator(AbstractContextManager):
```
Why do we need to inherit from `AbstractContextManager`? 
You use composition relationship you with the `ModelMonitoringSchedulesFile` that already inherits `AbstractContextManager` no?

jond01 replied:
Although the inheritance is not a must, it reminds us to use these classes as context managers when working with the underlying persistent data. It applies to both classes.
I get it might be irregular to see this. Do you have a suggestion?

davesh0812 commented on mlrun/model_monitoring/controller.py:
```diff
@@ -491,18 +457,28 @@ def _push_to_applications(
     :param context: the Nuclio context
     :param event:   trigger event
     """
+    context.platform.set_termination_callback(callback=_stop_running)
```
Can we run it only once when initializing the pod? Using `init_context(context)` facility of nuclio?

davesh0812 commented on mlrun/model_monitoring/db/_schedules.py:
```diff
@@ -0,0 +1,113 @@
+        self._fs = cast(fsspec.AbstractFileSystem, self._item.store.filesystem)
+        # `self._schedules` is an in-memory copy of the DB for all the applications for
+        # the same model endpoint.
+        self._schedules: dict[str, int] = {}
```
Needs to align with the actual inmemory data (self._item)

jond01 replied:
It's done in the context manager's `__enter__` method or `_open` method.
The constructor doesn't do IO intentionally, as it's unexpected and the reading is not always required - for example when creating a new file or deleting an existing file.

jond01 replied:
Added a flag and checks in 39e41ff 🙂

davesh0812 commented on mlrun/model_monitoring/db/_schedules.py:
```diff
@@ -0,0 +1,113 @@
+            )
+
+    def _open(self) -> None:
+        self._schedules = json.loads(self._item.get().decode(encoding=self.ENCODING))
```
what happens if somehow there is no file created?

jond01 replied:
Depending on the store, an error will be raised - `mlrun.errors.MLRunNotFoundError` or `BotoClientError`.

jond01 replied:
Added try-except and logs in 935e89d 👍

davesh0812 reviewed (commented):
Looks Good!
Few comments and questions :)
It seems that a few tests related to Windows machine updates were removed. Could we add those tests back under test_schedules.py?

jond01 committed:
Merge branch 'development' into task/controller-last-analyzed-in-json

jond01 committed:
Move the termination callback set to `init_context`

jond01 committed:
Add removed SQL tests to schedules tests

jond01 committed:
Fix

jond01 committed:
Merge branch 'development' into task/controller-last-analyzed-in-json

jond01 committed:
Add `_open_schedules` flag and checks

jond01 committed:
Improve the logs when the schedules file is not found

jond01 committed:
Merge branch 'development' into task/controller-last-analyzed-in-json

jond01 committed:
Improve in-memory store support

jond01 committed:
Remove the termination callback

It has no effect since Nuclio is blocked when it processes
a previous event.
https://iguazio.slack.com/archives/C5NB9CF0B/p1730379175509209

davesh0812 reviewed (approved):
Very Nice!!
